plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -2409,27 +2438,62 @@ read_only::get_table_by_scope_result read_only::get_table_by_scope( const read_o
    if( upper_bound_lookup_tuple < lower_bound_lookup_tuple )
       return result;
 
-   auto walk_table_range = [&]( auto itr, auto end_itr ) {
-      auto cur_time = fc::time_point::now();
-      auto end_time = cur_time + fc::microseconds(1000 * 10); /// 10ms max time
-      for( unsigned int count = 0; cur_time <= end_time && count < p.limit && itr != end_itr; ++itr, cur_time = fc::time_point::now() ) {
-         if( p.table && itr->table != p.table ) continue;
+   const bool reverse = p.reverse && *p.reverse;
+   const auto db_backing_store = get_backing_store();
+   if (db_backing_store == eosio::chain::backing_store_type::CHAINBASE) {
+      ilog("REM get_table_by_scope CHAINBASE");
+      auto walk_table_range = [&result,&p]( auto itr, auto end_itr ) {
+         keep_processing kp;
+         for( unsigned int count = 0; kp() && count < p.limit && itr != end_itr; ++itr ) {
+            if( p.table && itr->table != p.table ) continue;
 
-         result.rows.push_back( {itr->code, itr->scope, itr->table, itr->payer, itr->count} );
+            result.rows.push_back( {itr->code, itr->scope, itr->table, itr->payer, itr->count} );
 
-         ++count;
+            ++count;
+         }
+         if( itr != end_itr ) {
+            result.more = itr->scope.to_string();
+         }
+      };
+
+      const auto& d = db.db();
+      const auto& idx = d.get_index<chain::table_id_multi_index, chain::by_code_scope_table>();
+      auto lower = idx.lower_bound( lower_bound_lookup_tuple );
+      auto upper = idx.upper_bound( upper_bound_lookup_tuple );
+      if( reverse ) {
+         walk_table_range( boost::make_reverse_iterator(upper), boost::make_reverse_iterator(lower) );
+      } else {
+         walk_table_range( lower, upper );
       }
-      if( itr != end_itr ) {
-         result.more = itr->scope.to_string();
+   }
+   else {
+      ilog("REM get_table_by_scope ROCKSDB");
+      using namespace eosio::chain;
+      EOS_ASSERT(db_backing_store == backing_store_type::ROCKSDB,
+                 chain::contract_table_query_exception,
+                 "Support for configured backing_store has not been added to get_table_by_scope");
+      auto& kv_database = const_cast<chain::controller&>(db).kv_db();
```

Comment:
All these `const_cast` for kv database access need to be removed. Fix the interface of controller and combined_database.

Reply:
Fixed in both chain_plugin.cpp and chain_plugin.hpp.